src/analysis/processing/qgsalgorithmdistancewithin.cpp

```diff
@@ -39,28 +39,52 @@ void QgsDistanceWithinAlgorithm::process( const QgsProcessingContext &context, Q
     bool onlyRequireTargetIds,
     QgsProcessingFeedback *feedback, QgsExpressionContext &expressionContext )
 {
-  const bool isDynamicDistance = distanceProperty.isActive();
+  // By default we will iterate over the reference source and match back
+  // to the target source. We do this on the assumption that the most common
+  // use case is joining a points layer to a polygon layer (e.g. findings
+  // points near a polygon), so by iterating
+  // over the polygons we can take advantage of prepared geometries for
+  // the spatial relationship test.
+  bool iterateOverTarget = false;
+
+  //
+  // Possible reasons to iterate over target are considered here
+  //
+  do
+  {
+    // reference needs reprojection, so we cannot iterate over target
+    if ( targetSource->sourceCrs() != referenceSource->sourceCrs() ) break;
```

Comment:
This logic has changed -- if the target crs != reference crs then we have to iterate over the target instead.